### PR TITLE
Column support for array concat

### PR DESCRIPTION
**Pull request by jayzhan211**
# Which issue does this PR close?

Ref #6804

# Rationale for this change

Array concat with columns! Behavior is similar to Postgres.

Concat Operator `a || b` is not included in this PR!

# What changes are included in this PR?

# Are these changes tested?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 1]
```diff
@@ -389,6 +406,115 @@ select array_concat(make_array(10, 20), make_array([30, 40]), make_array([[50, 6
...
+# TODO: NULL should be empty not 0
+# query ?
+# select array_concat(make_array(column2), make_array(0)) from arrays_values;
+# ----
+# [1, 0]
+# [12, 0]
+# [23, 0]
+# [34, 0]
+# [44, 0]
+# [0, 0]
```
Comment (jayzhan211) [inline datafusion/core/tests/sqllogictests/test_files/array.slt:471]: `make_array` on the column with null values returns 0 but should be a null value instead. We can fix this in another PR.

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 2]
```diff
@@ -389,6 +406,115 @@ select array_concat(make_array(10, 20), make_array([30, 40]), make_array([[50, 6
...
+# TODO: " is missing
+# query ???
+# select array_concat(column1, column1), array_concat(column2, column2), array_concat(column3, column3) from arrays;
+# ----
+# [[, 2], [3, ], [, 2], [3, ]] [1.1, 2.2, 3.3, 1.1, 2.2, 3.3] [L, o, r, e, m, L, o, r, e, m]
+# [[3, 4], [5, 6], [3, 4], [5, 6]] [, 5.5, 6.6, , 5.5, 6.6] [i, p, , u, m, i, p, , u, m]
+# [[5, 6], [7, 8], [5, 6], [7, 8]] [7.7, 8.8, 9.9, 7.7, 8.8, 9.9] [d, , l, o, r, d, , l, o, r]
+# [[7, ], [9, 10], [7, ], [9, 10]] [10.1, , 12.2, 10.1, , 12.2] [s, i, t, s, i, t]
+# NULL [13.3, 14.4, 15.5, 13.3, 14.4, 15.5] [a, m, e, t, a, m, e, t]
+# [[11, 12], [13, 14], [11, 12], [13, 14]] NULL [,, ,]
```
Comment (jayzhan211) [inline datafusion/core/tests/sqllogictests/test_files/array.slt:484]: Not sure do we need an additional `""` for commas.

**Comment by jayzhan211** [comment]
Not all kinds of tests are passed but I think they can be solved in another PR. I think we can review this PR first.

File: datafusion/core/tests/sqllogictests/test_files/array.slt [hunk 3]
```diff
@@ -389,6 +410,150 @@ select array_concat(make_array(10, 20), make_array([30, 40]), make_array([[50, 6
...
+[L, o, r, e, m, ., ., .]
+[i, p, , u, m, ., ., .]
+[d, , l, o, r, ., ., .]
+[s, i, t, ., ., .]
+[a, m, e, t, ., ., .]
+[,, ., ., .]
+[., ., .]
+
+query ??I?
+select column1, column2, column3, column4 from arrays_values_v2;
```
Comment (jayzhan211) [inline datafusion/core/tests/sqllogictests/test_files/array.slt:515]: To remove

File: datafusion/core/tests/sqllogictests/test_files/array.slt (hunk 1 above)
Comment (jayzhan211) [reply datafusion/core/tests/sqllogictests/test_files/array.slt]: This is solved by https://github.com/apache/arrow-datafusion/pull/6900!

**Comment by jayzhan211** [comment]
@izveigor @alamb Ready for review! Thanks!

File: datafusion/physical-expr/src/array_expressions.rs [hunk 4]
```diff
@@ -554,42 +555,65 @@ fn align_array_dimensions(args: Vec<ArrayRef>) -> Result<Vec<ArrayRef>> {
...
+    let offsets: Vec<i32> = std::iter::once(0)
+        .chain(array_lens.iter().scan(0, |state, &x| {
+            *state += x;
+            Some(*state)
+        }))
+        .collect();
+
+    let builder = mutable.into_builder();
+
+    let list = builder
```
[the remark this answers is not in the document]
Comment (jayzhan211) [reply datafusion/physical-expr/src/array_expressions.rs:601]: I prefer to construct with ListArray::new instead of MutableArray too, but I fail on this.

**Review by alamb — APPROVED** [review]
Thank you @izveigor  and @jayzhan211

**Pull request merged**